secp256k1-sys/src/lib.rs

```diff
@@ -854,6 +864,10 @@ pub unsafe fn secp256k1_context_destroy(ctx: *mut Context) {
 ///
 /// See also secp256k1_default_error_callback_fn.
 ///
+///
+/// # Safety
+///
+/// For safety constraints see [`std::slice::from_raw_parts`] and [`std::str::from_utf8_unchecked`].
```

Comment:
In 6e19635acdf4e328748338bf47239ad41e7fb2de:

I don't understand what `std::str::from_utf8_unchecked` has to do with anything here. `slice::from_raw_parts` is sorta relevant but it's really overkill because it's talking in terms of an anbitrary `T` but we're using C characters, which are one byte in size, [have one-byte alignment](https://stackoverflow.com/questions/4637774/is-the-size-of-a-struct-required-to-be-an-exact-multiple-of-the-alignment-of-tha/4638295#4638295).

It would be clearer I think to just say that `message` must be a valid pointer to an initialized null-terminated C string.

Reply:
Ah, I see, the reason you're citing these is because in our default callback function we call `from_utf8_unchecked` after using `strlen` to compute a bound on the string in order to convert it to a u8 slice.

So we should add the additional requirement then that the string, up to the first null byte, must be valid UTF8.

Alternately we could call [`CStr::from_ptr`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_ptr) (and copy **its** safety bounds which are sensible for this funciton) and then [`CStr::to_str`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_str) which is safe but returns a Result, which we could just `.unwrap_or("<invalid UTF8>"` or something.

Reply:
Interestingly, doing this would eliminate our `strlen` funciton entirely, which is nice.

Reply:
Cool, that turned out nice. Pushed as a separate patch.